text/3808-register-tool.md

```diff
@@ -0,0 +1,287 @@
+
+Please do *not* use tool attributes for metadata that changes the meaning of the code. At that point you are parsing a dialect of Rust, and there is no indication for your users that their code will be interpreted differently by your tool than by the compiler.  For example, `#[must_use]` and `#[automatically_derived]` would be suitable for tool attributes, but `#[repr]` and `#[panic_handler]` are not, because they change the meaning of the code. For that use case, use proc-macros, generated code, or bare (un-namespaced) attributes instead, all of which will give a hard error if they cannot be understood by the compiler. If absolutely necessary to use bare attributes, use a C-style namespace like `#[rustc_const_stable]`.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Language
+
+### Background: name resolution of preludes
+
+Currently, names in the type namespace are resolved in [the following order](https://github.com/rust-lang/reference/pull/1765):
+
+1. Explicit definitions (including imports)
+2. [Extern prelude] (crates injected using either `extern crate` or `--extern`)
+3. [Tool prelude][builtin-tools]
+4. Standard library prelude
+5. Language prelude
+
+[Extern prelude]: https://doc.rust-lang.org/nightly/reference/names/preludes.html#extern-prelude
+
+Note that currently, loading a crate completely prevents using a tool attribute with that name. Consider this program:
+
+```rust
+extern crate rustfmt; // or --extern rustfmt
+
+#[rustfmt::skip]  // ERROR could not find `skip` in `rustfmt`
+fn foo ( ) { }
+```
+
+### Semantics
+
+The tool prelude is separated into the tool attribute prelude (which is in the type namespace) and the lint prelude (which is only active inside lint controls).
+
+The `#![register_attribute_tool(ident)]` crate-level attribute adds a new tool to the tool attribute prelude. The `#![register_lint_tool(ident)]` crate-level attribute adds a new tool to the lint prelude. For both attributes, tools must be a single ident, not a nested path.
+
+`register_attribute_tool` and `register_lint_tool` are idempotent; duplicating the attribute has no effect.
+
+Crate-level macros such as `#![cfg_attr(...)]` may expand to `register_*_tool`, subject to all existing rules for macro expansion.
+
+Like today, attributes and lints in a tool namespace are always considered used by the compiler. The compiler does not verify the contents of any tool attribute, except to verify that all attributes are syntactically valid [tool attributes].
+
+Registering a predefined tool (`clippy`, `miri`, etc.) using `#![register_*_tool(...)]` is an error.
```

Comment:
Where will this list be maintained?

Also, today you can't add any attribute that starts with `rustc`. Will that still be the case?

Reply:
From what I understand, today you cannot add a tool attribute that includes the prefix `rustc`, for example, you can register a tool `rustc_foo`, but the compiler will reject the attribute `#[rustc_foo::bar]`.

Reply:
oh, i see. this is actually not coming from the register_tool code, this is the same code that rejects `#[rustc_baz]`.

i think it is fairly reasonable to keep this reserved, but i can make sure it's rejected earlier in `register_tool` so that people aren't confused.

Reply:
>but i can make sure it's rejected earlier in register_tool so that people aren't confused.

I don't think it makes sense to prevent this, there are many other ways to make an unusable attribute that are not rejected, like `mod rustc_something { pub use my_attribute; }` or `use my_attribute as rustc_something;`, but we do not prohibit modules or imports containing `rustc` preventively.

Reply:
As long as we are consistent. Either it works without any hacks or it fails early on.